src/main/java/com/snowflake/kafka/connector/internal/SnowflakeInternalStage.java

```diff
@@ -65,72 +106,139 @@ public SnowflakeInternalStage(
-            (SnowflakeFileTransferMetadataV1) agent.getFileTransferMetadatas().get(0);
-        if (fileTransferMetadata.getStageInfo().getStageType() != StageInfo.StageType.GCS) {
-          // Overwrite the credential to be used
-          credential =
-              new SnowflakeMetadataWithExpiration(fileTransferMetadata, System.currentTimeMillis());
-          storageInfoCache.put(stage, credential);
-        } else {
-          throw SnowflakeErrors.ERROR_5017.getException();
-        }
-      }
+      SnowflakeMetadataWithExpiration credential = storageInfoCache.getOrDefault(stageName, null);
 
-      SnowflakeFileTransferMetadataV1 fileTransferMetadata = credential.fileTransferMetadata;
-      // Set filename to be uploaded
-      fileTransferMetadata.setPresignedUrlFileName(fullFilePath);
-
-      byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
-      InputStream inStream = new ByteArrayInputStream(dataBytes);
-
-      // This uploadWithoutConnection api cannot handle expired credentials very well.
-      // Need to prevent passing expired credential to it.
-      try {
-        SnowflakeFileTransferAgent.uploadWithoutConnection(
-            SnowflakeFileTransferConfig.Builder.newInstance()
-                .setSnowflakeFileTransferMetadata(fileTransferMetadata)
-                .setUploadStream(inStream)
-                .setRequireCompress(true)
-                .setOcspMode(OCSPMode.FAIL_OPEN)
-                .setProxyProperties(proxyProperties)
-                .build());
-      } catch (Throwable t) {
-        // If this api encounters error, invalid the cached credentials
-        // Caller will retry this function
-        logWarn("uploadWithoutConnection encountered an error for fileName:{}", fullFilePath);
-        storageInfoCache.remove(stage);
-        throw t;
+      if (!isCredentialValid(credential, stageType)) {
+        logDebug(
+            "Query credential(Refreshing Credentials) for stageName:{}, filePath:{}",
+            stageName,
+            fullFilePath);
+        refreshCredentials(stageName, stageType, fullFilePath);
       }
+    } catch (Exception e) {
+      logWarn(
+          "Failed to refresh Credentials for stageName:{}, filePath:{}", stageName, fullFilePath);
+      throw SnowflakeErrors.ERROR_5018.getException(e.getMessage());
+    }
+
+    // Lets fetch the file transfer metadata from cache again. (Because we might have refreshed
+    // them)
+    SnowflakeFileTransferMetadataV1 fileTransferMetadata =
+        storageInfoCache.get(stageName).fileTransferMetadata;
+    // Set filename to be uploaded
+    // This set is not useful in GCS since there is a bug in JDBC which doesnt use destFileName.
+    fileTransferMetadata.setPresignedUrlFileName(fullFilePath);
 
+    byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
+    InputStream inStream = new ByteArrayInputStream(dataBytes);
+
+    // This uploadWithoutConnection api cannot handle expired credentials very well.
```

Comment:
What happens if we pass an expired credential? Currently expiration buffer is 30 seconds, should we increase it to 5 minutes to be sure that this doesn't happen?

Reply:
Technically here it how I see this affecting customers:
- If we pass in expired credentials, we will get expired credentials error from cloud. 
- JDBC itself should renew the token but looks like that part of the code is not working. https://github.com/snowflakedb/snowflake-jdbc/blob/master/src/main/java/net/snowflake/client/jdbc/cloud/storage/SnowflakeS3Client.java#L738
- We should be doing a retry for this failure and renew the credentials from our end. 

Worst case we would put this file inside a table stage. (Which is considered as data loss)

Reply:
Thanks for the explanation, this makes sense. Can you add this in comment for future reference?
Also are we currently retrying for this failure? If not, we can decouple it from this work but we should track it

Reply:
Yes, lemme copy paste this in docs and we are retrying it. But we dont have enough logs to understand if retry is successful or not. (So added retry logging in this PR as well)

Also, logger level in this exceptions are warn and after max retries, it is error.